### PR TITLE
First pass mypy typing 

### PR DESCRIPTION
**Pull request by bryevdv**
This PR adds begins adding Python types Cunumeric, starting with the following modules
```
cunumeric._sphinxext
cunumeric._ufunc
cunumeric.coverage
cunumeric.fft
cunumeric.install_info
cunumeric.linalg
cunumeric.patch
cunumeric.random
cunumeric.utils
cunumeric.window
```
Note the ugly `mypy_path` hack currently in `pyproject.toml`:
```
mypy_path = "typings/:../legate.core/typings/:..legate.core/install38/lib/python3.8/site-packages/"
```
This is because `legate` is not actually installed in the `site-packages` for the Python that is executing `mypy`. I think probably the only solution is a dedicated script to invoke `mypy` and also set the env var `MYPY_PATH` from computed values for the legate install. But if anyone has any other ideas, I would love to hear them.

### CODE REVIEW

**Comment by bryevdv** [comment]
Working on `linalg` now and running into issues with the code that uses `legate.core` directly. I think actually we will have to simply check both together, in order to check `cunumeric`:
```
mypy ../legate.core/install38/lib/python3.8/site-packages/legate /home/bryan/work/legate.core/typings/ cunumeric
```
I tried several approaches with `MYPYPATH` and it was just not picking up imports and types correctly.

Anyway, I will stop this PR for now as a first step.

**Comment by bryevdv** [comment]
Once https://github.com/nv-legate/cunumeric/pull/387 is merged, then `linalg` also works, with either of these invocations:
```
~/work/cunumeric bryanv/mypy*
dev38 ❯ mypy ../legate.core/legate /home/bryan/work/legate.core/typings/ cunumeric
Success: no issues found in 79 source files

~/work/cunumeric bryanv/mypy*
dev38 ❯ mypy ../legate.core/install38/lib/python3.8/site-packages/legate /home/bryan/work/legate.core/typings/ cunumeric
Success: no issues found in 79 source files
```
I assume passing the installed version will be more CI-friendly? If so then I will add a script to wrap the `mypy` invocation to include this.

**Comment by manopapad** [comment]
> This is because legate is not actually installed in the site-packages for the Python that is executing mypy. I think probably the only solution is a dedicated script to invoke mypy and also set the env var MYPY_PATH from computed values for the legate install. But if anyone has any other ideas, I would love to hear them.

We could just bite the bullet and update our setup process to always install legate/cunumeric properly in the currently active virtual environment, instead of using a custom directory that needs to be passed everywhere. We will need to do this anyway eventually, as the `setuptools` deprecation warnings keep telling us.

**Comment by bryevdv** [comment]
> We could just bite the bullet and update our setup process to always install legate/cunumeric properly in the currently active virtual environment, instead of using a custom directory that needs to be passed everywhere. We will need to do this anyway eventually, as the setuptools deprecation warnings keep telling us.

I think that would be *wonderful* but also sounds like a bit of an endeavor. I guess I would not hold up this PR for that, since a wrapper script is a simple solution for the immediate term.

**Comment by bryevdv** [comment]
oops, introduced two minor issues. Should go green now (all tests pass locally eager/gpu/cpu)

**Comment by bryevdv** [comment]
Merge conflict is resolved.

File: cunumeric/fft/fft.py
Comment (bryevdv) [inline cunumeric/fft/fft.py:940]: Just noting I made similar changes throughout. In general, whenever there is a "convenience" arg that can accept a variety of types, to be internally adapted to single type, it's better to use a new variable to hold the computed value. (e.g. otherwise any functions that receive the value would have to accept the "wide" type specification, and/or mypy will need annoying explicit type annotations on the variable)

File: cunumeric/linalg/cholesky.py
Comment (bryevdv) [inline cunumeric/linalg/cholesky.py:45]: I don't think there is a way to narrow the type returned by `create_task`, so this cast is necessary. We could bifurcate `create_task` on the boolean parameter.

File: cunumeric/linalg/linalg.py
Comment (bryevdv) [inline cunumeric/linalg/linalg.py:441]: The very slight difference in this logic is that the check will also happen for the `axis=None` case, which is slightly redundant since it will always succeed in that case by design. I think reducing the nested complexity is worth the trade-off tho.

File: cunumeric/_ufunc/ufunc.py
Comment (magnatelee) [inline cunumeric/_ufunc/ufunc.py:181]: I feel we should remove `types` from this parent class, as `types` in each sub-class has a different type (`dict[str, str]` for unary_ufunc, `dict[str, tuple[str, str]]` for multiout_unary_ufunc, and `dict[tuple[str,str],str]` for binary_ufunc)

File: cunumeric/fft/fft.py
Comment (magnatelee) [reply cunumeric/fft/fft.py:940]: I'll make a note about this in the weekly meeting next week.

File: cunumeric/linalg/cholesky.py
Comment (magnatelee) [reply cunumeric/linalg/cholesky.py:45]: Let's bifurcate `create_task`. Otherwise, there will be two more ugly castings like this once #391 is merged.

File: cunumeric/linalg/cholesky.py
Comment (magnatelee) [inline cunumeric/linalg/cholesky.py:142]: I think we should change the body of this function to make the return type be `Shape`.

File: cunumeric/linalg/linalg.py
Comment (magnatelee) [inline cunumeric/linalg/linalg.py:23]: just curious, what's the reason for this comment?

File: cunumeric/linalg/linalg.py
Comment (magnatelee) [inline cunumeric/linalg/linalg.py:86]: I think the return type should be just `ndarray`. Add a `None` check if you need it. Note that the while loop at line 152 always runs at least once.

File: cunumeric/linalg/linalg.py
Comment (magnatelee) [inline cunumeric/linalg/linalg.py:301]: the type of `order` must be `npt.NDArray[np.int64]`. I think you need an `isinstance(order, npt.NDArray[np.int64])` at the callsite or need to remove `return_costs` from  function `_multi_dot_matrix_chain_order`, as I don't think that function is actually used with `return_costs=True`. (If you decide to do the latter, then the return type of that function will also be simplified.)

File: cunumeric/random/random.py
Comment (magnatelee) [reply cunumeric/random/random.py:63]: I agree that this was a bug.

File: cunumeric/_ufunc/__init__.py
Comment (bryevdv) [reply cunumeric/_ufunc/__init__.py:15]: We don't *need* it in any empty module (which could be an init, or not). However I have a strong preference for these to be pure boilerplate, so that no thinking is required (and no future issues in case something is added to the module in the future)

File: cunumeric/_ufunc/ufunc.py
Comment (bryevdv) [reply cunumeric/_ufunc/ufunc.py]: https://github.com/nv-legate/cunumeric/pull/387/commits/f70ae583d855ad9d2f9707e7b8fd4c44be9bf6d2

File: cunumeric/linalg/linalg.py
Comment (bryevdv) [reply cunumeric/linalg/linalg.py]: https://github.com/nv-legate/cunumeric/pull/387/commits/95b3c15f513150053b39a277904816ec562daf2d

File: cunumeric/linalg/linalg.py
Comment (bryevdv) [reply cunumeric/linalg/linalg.py]: https://github.com/nv-legate/cunumeric/pull/387/commits/fdf2eee4b2a451ff3aa997d144fa6a3b9be40dc4

File: cunumeric/_ufunc/ufunc.py
Comment (bryevdv) [reply cunumeric/_ufunc/ufunc.py:229]: OK, I will make `create_auto_task` and `create_manual_task` that delegate to `create_task` but since that will require `legate` changes I propose to merge this for now, and then make update in the next mypy PR once the legate change is also merged.

File: cunumeric/linalg/cholesky.py
Comment (bryevdv) [reply cunumeric/linalg/cholesky.py:62]: OK, I will make create_auto_task and create_manual_task that delegate to create_task but since that will require legate changes I propose to merge this for now, and then make update in the next mypy PR once the legate change is also merged.

File: cunumeric/_ufunc/ufunc.py
Comment (bryevdv) [reply cunumeric/_ufunc/ufunc.py:270]: Sure, there is a lot of overloaded usage of "shape" I'd prefer to make a comprehensive changes for convenience types like this across both repos after the initial document-current-status work is done.

File: cunumeric/linalg/cholesky.py
Comment (bryevdv) [inline cunumeric/linalg/cholesky.py:128]: Diff here was simply removing the superfluous `else`.

**Review by magnatelee — APPROVED** [review]

**Pull request merged**
